sys/linux: alignment for integer ranges

pchaigno opened the pull request:
This pull request enables the syntax `intN[start:end, alignment]` for integer ranges, as discussed in #1452. For instance, `int32[1:11, 2]` represents odd 32-bit numbers between 1 and 11 included. With this change, two `NEED` tags in syscall descriptions can be addressed.

I would like to address a third `NEED` tag in `sys/linux/dev_udmabuf.txt`:
```diff
- # NEED: this should be a page-aligned integer
- udmabuf_offset_size = 0, 0x1000, 0x2000, 0x4000, 0x8000, 0x10000, 0x1000000, 0x100000000, 0xfffff000, 0x1000000000000, 0xfffffffffffff000, 0xfffffffff0000000
+ int64[0:0xffffffffffffffff, 0x1000]
```
Using such a large value for the end of the range is however not possible since 710eefe85a976c438da255499fbefd1a6c989ef6. One solution could be to define a special value to mean the "all possible values", such as `int64[0:0, 0x1000]`, but that may not be the preferred solution?

EDIT: I've broken the pull request in two distinct commits (support then use).

dvyukov commented on prog/rand.go:
```diff
@@ -119,10 +119,14 @@ func truncateToBitSize(v, bitSize uint64) uint64 {
 		return r.randInt(bitSize)
 	}
+	if align != 0 {
+		endAlign := (end - begin) / align
```
This needs to be (end - begin + 1).

pchaigno replied:
Do you have an example where the above doesn't work? If I do `+ 1`, `[1:100, 50]` doesn't return the expected values: `[1:100, 50] -> 1 + [0:2] * 50 -> {1, 51, 101}`.

dvyukov replied:
You are right.

dvyukov commented on pkg/compiler/types.go:
```diff
@@ -83,13 +86,16 @@ var typeInt = &typeDesc{
 				rangeEnd = rangeArg.Colon[0].Value
 			}
+			if len(args) > 1 {
+				align = args[1].Value
```
This needs some checks. I can think of at least [1:7, 10] that does not make sense.

dvyukov replied:
Also align 0 or 1 or negative don't make sense.

pchaigno replied:
Okay. I wasn't sure you'd want to enforce this since `align` could be a constant and sometimes be greater than `rangeEnd`, sometimes lower. `[1:7, 10]` would still return `1`.

dvyukov replied:
But what's the point of [1:7, 10]? That's just const[1]? Frequently such things mean a typo/bug in descriptions. And since we don't have any real testing for the descriptions, usually such warnings find a number of real bugs.

pchaigno replied:
> But what's the point of [1:7, 10]? That's just const[1]?

Not much point. I just wanted to allow for `[1:7, SOME_CONSTANT_FROM_KERNEL]`, but that's probably a corner case and finding bugs in descriptions is more important. I'll add checks!

dvyukov replied:
If SOME_CONSTANT_FROM_KERNEL does not vary across archs (or there is a corresponding literal constant for range end), then it's still possible to do, right?
Also may be solvable with defines:
```
define RANGE_END SOME_CONSTANT_FROM_KERNEL * 10
```
Let's first hit such case, relaxing checks is always possible.

pchaigno replied:
Okay, that makes sense.

pchaigno replied:
I've rebased with an updated patchset. Only subranges of `[-MAX_INTN:MAX_INTN]` for any `intN` are allowed. Is that what you had in mind? I wasn't sure how to treat negative values.

dvyukov commented on pkg/compiler/testdata/all.txt:
```diff
@@ -12,6 +12,7 @@ foo$7(a int8[-20:20])
 foo$8(a ptr[in, strings])
 foo$9(a ptr[out, ptr[in, string]])
 foo$10(a ptr[out, buffer[in]])
+foo$11(a int64[1:100, 2])
```
We also need some error tests for bad align value, non-number, etc.

dvyukov commented:
> One solution could be to define a special value to mean the "all possible values", such as int64[0:0, 0x1000]

Maybe we could non error on int64[0:-1] and generate non-ranged int from this. But only if alignment is present, b/c otherwise it does not make sense.

dvyukov commented:
2 unrelated things that come to mind while looking at this:
1. Check range values and const value if they fit into the base type, e.g. `int8[1000]` is a bug.
2. Need to add NEED: comment for power-of-2 int's, there seems to be much more of them than aligned ints (I've grepped for "1, 2, 4").
Thought maybe you are up to improving that as well :)

pchaigno commented on prog/rand.go:
```diff
@@ -119,10 +119,14 @@ func truncateToBitSize(v, bitSize uint64) uint64 {
 	if r.oneOf(100) {
 		return r.randInt(bitSize)
 	}
+	if align != 0 {
```
Hm. This block should be before the previous `r.oneOf(100)` block. Otherwise, we risk overflowing and we're doubling the chances of calling `randInt()`. I'll fix.

dvyukov replied:
But if we move it above, we will never generate unaligned value, right? I think we want unaligned once in a while too.

pchaigno replied:
Yep, you're right, and I was wrong regarding overflows, they'll still happen. I'm guessing overflows on generated values are not really a concern though (?)
Is doubling the chances of calling `randInt()` when we have an alignment parameter an issue? If not, we can leave it that way. If it is, I can rewrite to remove the recursive call.

dvyukov replied:
I would say it's fine. Probability of generating completely random values is quite low. I can also see some sense in generating completely random but aligned values (which we will get in case of recursion).

pchaigno replied:
Okay.

dvyukov commented:
Also need a change to IntType.getMutationPrio -- the aligned int may have significantly fewer values.

pchaigno commented on docs/syscall_descriptions_syntax.md:
```diff
@@ -95,7 +96,8 @@ example_struct {
 	f1	const[0x42, int16be]	# const 2-byte integer with value 0x4200 (big-endian 0x42)
 	f2	int32[0:100]		# random 4-byte integer with values from 0 to 100 inclusive
-	f3	int64:20		# random 20-bit bitfield
+	f3	int32[1:10, 2]		# random 4-byte integer with values {1, 3, 5, 7, 9}
```
@dvyukov Have you seen that example?
I'm wondering if we're on the same page regarding the semantics of the alignment. I've implemented it to be an alignment starting from `beginRange`, not from 0. Your other comments seem to assume its an alignment to 0.
Aligning on `beginRange` seemed more generic; for instance, it allows us to express odd values as above. It may need to be said explicitly in the description though.

dvyukov replied:
I agree. It should be alignment from range begin.

pchaigno commented:
> Maybe we could non error on int64[0:-1] and generate non-ranged int from this. But only if alignment is present, b/c otherwise it does not make sense.

Okay, that looks fine.

> Need to add NEED: comment for power-of-2 int's, there seems to be much more of them than aligned ints (I've grepped for "1, 2, 4").
Thought maybe you are up to improving that as well :)

It's probably best if we keep that for another PR. I'm still making a lot of small mistakes in pull requests, so I'd rather keep them small for now.

Although, we should probably discuss how that would look like so that we don't need to go back on the alignment syntax later. Are you thinking just an additional argument? E.g., `intN[start:end, align, power]`, with `align=1` allowed when we introduce `power`.

> Also need a change to IntType.getMutationPrio -- the aligned int may have significantly fewer values.

Ah, right. Forgot that one; will fix.

dvyukov commented:
> It's probably best if we keep that for another PR.

Yes. I just needed to record it somewhere, so that it's not lost...

> Although, we should probably discuss how that would look like

I meant just adding NEED marks at least for some of them. But if you want to attack it, that's even better. But yes, then we need to figure out how to do it first... good question. Observation 1: power is mutually exclusive with align, right? Observation 2: we probably only need power-of-2.
One option would be what you proposed, e.g. `int32[1:32, 1, 2]`.
Another that I can think would be `int32[1:32, -2]` with a special meaning for negative alignment (theoretically extendable to any other power, but I think we should restrict it only to -2 for now).

pchaigno commented:
Okay. I just wanted to make sure it wouldn't clash with the align syntax in any way. It won't, so we can leave it to a later pull request. Thanks!

pchaigno commented on pkg/compiler/types.go:
```diff
@@ -81,15 +84,44 @@ var typeInt = &typeDesc{
+			}
+			size, _ := comp.parseIntType(t.Ident)
+			size = size * 8
+			if size < 64 && (end+1<<size > 2*(1<<size)-1 || begin+1<<size > 2*(1<<size)-1) {
```
To be honest, I'm a little surprised the above works and the compiler doesn't try to optimize it. The idea is we want values in `[0:MAX_INTN]` or `[MAX_INT64-MAX_INTN:MAX_INT64]`. So we first shift all values by `MAX_INTN+1` and check bounds afterward.

pchaigno replied:
It's not totally clear to me what values make and don't make sense. I currently allow `intN[-MAX_UINTN:MAX_UINTN]`, but maybe we should only allow `intN[-MIN_SINTN:MAX_SINTN]` and `intN[0:MAX_UINTN]` (as I understand, `int32` can either mean `int32_t` or `uint32_t`)?

pchaigno replied:
I implemented a check for the above, which therefore prohibits `int8[-200:200]`, but also `int8[-200:0]`.

pchaigno commented on sys/linux/dev_infiniband_rdma_cm.txt:
```diff
@@ -132,7 +132,7 @@ rdma_ucm_accept {
 
 rdma_ucm_reject {
 	id			rdma_cm_id
-	private_data_len	int8[0:RDMA_MAX_PRIVATE_DATA]
```
`RDMA_MAX_PRIVATE_DATA` is 256 so we're actually asking for all possible `int8` values here. I'm not sure what the intent was here, so not sure I've got the correct fix. Should `private_data_len` be the bytesize of `private_data`?

dvyukov commented on pkg/compiler/testdata/all.txt:
```diff
@@ -14,6 +14,7 @@ foo$9(a ptr[out, ptr[in, string]])
 foo$10(a ptr[out, buffer[in]])
 foo$11(a int64[1:100, 2])
 foo$12(a int64[0:-1, 0x1000])
+foo$13(a int8[-255:255])
```
Please add more tests, they are cheap. For begin/end separately, for bitfields.

pchaigno replied:
Ah, bitfields. Forgot about these. I'll add more tests for bitfields and will fix the bitsize in `IntType.CheckConsts`.

dvyukov commented on pkg/compiler/types.go:
```diff
@@ -81,15 +84,37 @@ var typeInt = &typeDesc{
 		typeArgBase.Type.Check(comp, t)
 	},
+	CheckConsts: func(comp *compiler, t *ast.Type, args []*ast.Type, base prog.IntTypeCommon) {
+		if len(args) > 0 && len(args[0].Colon) != 0 {
```
What if we have int[100, 200], it won't pass any verification?

pchaigno replied:
That's allowed? Won't it be a `const`? I couldn't find any constant expressed as an `int`, so I assumed it wasn't allowed by the syntax :/

dvyukov replied:
It should not be allowed. But as far as I see the new code won't reject it b/c len(args[0].Colon) == 0 in this case, so we will not do any checks whatsoever. We should.

pchaigno replied:
I added a check that the first int argument always is a range, i.e., contains a colon.

dvyukov commented on pkg/compiler/types.go:
```diff
@@ -81,15 +84,37 @@ var typeInt = &typeDesc{
+				comp.error(args[0].Pos, "bad int range [%v:%v]", begin, end)
+				return
+			}
+			if len(args) > 1 && args[1].Value > end {
```
In randRangeInt we do `endAlign := (end - begin) / align`. So I would check that we get endAlign at least not 0 (for int32[9:10, 8]).

pchaigno replied:
Fixed with the last update.

dvyukov commented on pkg/compiler/types.go:
```diff
@@ -81,15 +84,37 @@ var typeInt = &typeDesc{
+		if len(args) > 0 && len(args[0].Colon) != 0 {
+			begin := args[0].Value
+			end := args[0].Colon[0].Value
+			if len(args) > 1 && begin == 0 && int64(end) == -1 {
```
This only makes sense if we have align, right?
Otherwise I think we will get division by 0 in `return begin + (r.Uint64() % (end - begin + 1))`.

pchaigno replied:
Yes. the `len(args) > 1` condition should ensure we only allow these values when we have an `align` argument... or have I missed something?

dvyukov replied:
Right.

dvyukov commented on prog/rand.go:
```diff
@@ -119,10 +119,18 @@ func truncateToBitSize(v, bitSize uint64) uint64 {
+	if align != 0 {
+		if begin == 0 && int64(end) == -1 {
+			// Special [0:-1] range for all possible values.
+			end = uint64(math.Exp2(float64(bitSize)) - 1)
```
Isn't it the same as `1<<bitSize - 1`? I am afraid of floating points.

pchaigno replied:
It is. Fixed with the last update.

dvyukov commented on prog/mutation.go:
```diff
@@ -501,7 +501,11 @@ func (t *IntType) getMutationPrio(target *Target, arg Arg, ignoreSpecial bool) (
 
-	switch size := t.RangeEnd - t.RangeBegin + 1; {
+	size := t.RangeEnd - t.RangeBegin + 1
+	if t.Align != 0 {
```
We may also need changes to mutateInt because it tries to add/subtract small values every other time.

pchaigno replied:
Done in the last update.

dvyukov commented on prog/mutation.go:
```diff
@@ -501,7 +501,11 @@ func (t *IntType) getMutationPrio(target *Target, arg Arg, ignoreSpecial bool) (
-	switch size := t.RangeEnd - t.RangeBegin + 1; {
+	size := t.RangeEnd - t.RangeBegin + 1
+	if t.Align != 0 {
+		size = (t.RangeEnd-t.RangeBegin)/t.Align + 1
```
For the [0:-1] special case we will into the plainPrio below, right? I think that's right behavior.

pchaigno replied:
Actually, that's an oversight. Do we really want to go into plainPrio if we have e.g., `int8[0:-1, 50]`? That's less than 6 values, so I'd have thought we wanted to go into `size <= 15`.

pchaigno commented:
I've updated the branch with several changes:
- Check for ranges w.r.t. base type's size now checks that range is within `intN[-MIN_SINTN:MAX_SINTN]` or `intN[0:MAX_UINTN]`. `int8[-200:0]` is therefore prohibited.
- Added more tests for bitfields and error cases.
- Added a check that first argument of `intN` is a range (has a colon). Fixed a few `intN` that should be `const` in syscall descriptions.
- Updated `IntType.mutate()` to mutate aligned int "properly". Currently, the value is changed in the same way as a "normal int" but kept aligned. The value therefore increments/decrements by a `n * align`.
- Updated `getMutationPrio()` to handle special `[0:-1]` case.
- Rebased.

There are probably more commits than necessary because I tried to keep the different changes (check first int arg is range, alignment support, and check range consistency with base type) and corresponding changes to syscall descriptions separated. I can squash commits as preferred if needed of course.
